### ISSUE TITLE
Bootstrap Service Encryption using Secret Encryption Key not working

### ISSUE THREAD

**Issue by wmkamerman** [absmach/magistrala#1450]
**BUG REPORT**

1. What were you trying to achieve?
I am trying to use the Bootstrap service over HTTP instead of HTTPS because we have trouble sending HTTPS requests with our device. We CAN get MQTT working over TLS so I want to use the bootstrap service to retrieve Thing ID, Key, Channels and the required certificates. To ensure safe communication, I am trying to use the Secret encryption key as mentioned in the Docs:
> Both device and Bootstrap service use a secret key to encrypt the content. Encryption is done as follows:
> 
> 1) Device uses the secret encryption key to encrypt the value of that exact external key 2) Device sends a bootstrap request using the value from 1 as an Authorization header 3) Bootstrap service fetches config by its external ID 4) Bootstrap service uses the secret encryption key to decrypt Authorization header 5) Bootstrap service compares value from 4 with the external key of the config from 3 and proceeds to 6 if they're equal 6) Bootstrap service uses the secret encryption key to encrypt the content of the bootstrap response
> 
> Please have on mind that secret key is passed to the Bootstrap service as an environment variable. As security measurement, Bootstrap service removes this variable once it reads it on startup. However, depending on your deployment, this variable can still be visible as a part of your configuration or terminal emulator environment.
> 
> For more details on which encryption mechanisms are used, please take a look at the implementation.

2. What are the expected results?
I added a Bootstrap configuration and sent a GET request with my External ID and External Key to the following URL:
http://host:8202/things/bootstrap/:externalId
Header: "Authorization: <external_key>"

The Bootstrap service then returns my configuration as expected. All goes well! 

Now I try to get the encrypted version working. According to the docs, I should encrypt the External Key using the predefined secret. I inspected the code and from what I can tell, there is some AES encryption. My secret key is the default value: 12345678910111213141516171819202, which I have passed as MF_BOOTSTRAP_ENCRYPT_KEY environment variable.

I open the following URL and entered the secret key as secret and my external key as text to be encrypted. https://www.devglan.com/online-tools/aes-encryption-decryption 

I then enter the output value as Authorization header instead of the non-encrypted External Key. I then expect a Bootstrap response that is encrypted using the same secret. The device can then decrypt.

3. What are the received results?
With both CBC and ECB modes and both BASE64 and HEX output used as Authorization header, I get the error: failed to get bootstrap configuration for given external key : non-existent entity

4. What are the steps to reproduce the issue?
They are described above.

5. In what environment did you encounter the issue?
A Docker compose environment, with the MF_BOOTSTRAP_ENCRYPT_KEY passed to the Docker container as environment variable.

6. Additional information you deem important:
None specifically. I love that you have thought about both HTTP and HTTPS communication and the concept of this encryption method!

**ENHANCEMENT**
1. Describe the enhancement you are requesting.
Improved documentation on how to use the encryption feature. There is some information lacking, as the docs mention only:

> For more details on which encryption mechanisms are used, please take a look at the implementation.

2. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
Should-have

**Comment by wmkamerman** [comment]
Update: I have discovered the things I do wrong.

First, there is a separate endpoint for the encrypted communication
`/things/bootstrap/secure/{externalId}`
https://api.mainflux.io/?urls.primaryName=bootstrap.yml#/configs/get_things_bootstrap_secure__externalId_

Note the extra `/secure/` in the URL, and the fact that you need to send an Authorization header and not configAuthorization as mentioned in the API docs.

Then we get to the encryption key. Choose a secret key with length 16, 24 or 32, which defines whether AES-128, AES-192 or AES-256 is used. Hex encode the secret key and pass the outcome to the MF_BOOTSTRAP_ENCRYPT_KEY environment variable.

Next, instead of sending the unencrypted 'external_key' as Authorization header to the endpoint, you need to encrypt the external_key using the secret key. I managed to get it working by running some code I found in the repo. More info will follow to make it clear for everyone, as you need to include some IV which is set to 16 random bytes by default. Important to note is that Mainflux uses CFB encryption. We use a 16 char secret key, thus AES-128. Probably that is why we also need to encrypt using CFB-128.

Now that the endpoint correctly functions, we can decode the result by using the secret key, as well as the IV (Initialization Vector). The IV is equal to the first 16 bytes of the binary response. Then decode the _rest_ of the response (everything after the first 16 bytes) using the CFB-128 algorithm.

Long story short, the endpoint does work as expected, but the documentation is lacking lots of necessary information on this topic!

**Comment by nmarcetic** [comment]
@wmkamerman Glad, you solve it. 
I agree that documentation doesn't cover user stories and different scenarios, it's really hard to do that. 

We wrote a blog with [complete example](https://medium.com/mainflux-iot-platform/mainflux-bootstrap-service-a140d5f4c60d)

All contributions are more than welcome if you feel comfortable helping by improving documentation.
